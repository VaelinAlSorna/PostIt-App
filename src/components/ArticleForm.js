import React from 'react';
import moment from 'moment';

class ArticleForm extends React.Component {
    state = {
        title: '',
        text: '',
        createdAt: moment().format('D. M. Y')
    };
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({
            title
        }))
    }
    onTextChange = (e) => {
        const text = e.target.value;
        this.setState(() => ({
            text
        }))
    }
    onSubmit = (e) => {
        e.preventDefault();
        this.props.onSubmit(this.state)
    }
    render () {
       return ( 
        <div> 
            <form onSubmit={this.onSubmit}> 
                <input 
                    onChange={this.onTitleChange} 
                    placeholder="Title" 
                    type="text" 
                    value={this.state.title}
                />
                <textarea 
                    onChange={this.onTextChange}
                    placeholder="Add text for your article"
                    value={this.state.text}
                >
                </textarea>
                <input type="submit" value="Submit" />
            </form>
        </div>
       );
    }
}

export default ArticleForm;