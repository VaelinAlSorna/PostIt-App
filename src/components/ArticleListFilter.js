import React from 'react';
import { connect } from 'react-redux';
import { setYourArticlesFilter, removeYourArticlesFilter, setTextFilter } from '../actions/filterActions';

export class ArticleListFilter extends React.Component {
    handleSeeAll = () => {
            this.props.removeYourArticlesFilter();
            this.props.setTextFilter('');
    }
    handleSeeYours = () => {
        this.props.setYourArticlesFilter();
    }
    disablingBtnSeeAll = () => {
        if(this.props.filter.yourArticles || !!this.props.filter.text) {
            return false;
        } else {
            return true;
        }
    }
    disablingBtnSeeYours = () => {
        if(this.props.user.name && !this.props.filter.yourArticles) {
            return false;
        } else {
            return true;
        }
    }
    onTextFilterChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }
    onSubmit = (e) => {
        //This is here in order to prevent user on mobile from sunmiting form
        e.preventDefault();
    }
    
    render () { 
        return  ( 
            <div className="filter">
                <form onSubmit={this.onSubmit} className="filter__group filter-form"> 
                    <input
                        className="filter-form__input"
                        onChange={this.onTextFilterChange} 
                        type="text" 
                        placeholder="Search Article By Name"
                        value={this.props.filter.text}
                    />
                </form>
                <button className="filter__group btn btn--green" onClick={this.handleSeeAll} disabled={this.disablingBtnSeeAll()}>See All Articles</button>
                <button className="filter__group btn btn--green" onClick={this.handleSeeYours} disabled={this.disablingBtnSeeYours()}>See Your Articles</button>
            </div>                
        )
    };
}

const mapStateToProps = (state) => ({
    filter: state.filter,
    user: state.auth
});

const mapDispatchToProps = (dispatch) => ({
    setYourArticlesFilter: () => dispatch(setYourArticlesFilter()),
    removeYourArticlesFilter: () => dispatch(removeYourArticlesFilter()),
    setTextFilter: (text) => dispatch(setTextFilter(text))
});

export default connect(mapStateToProps, mapDispatchToProps)(ArticleListFilter);

