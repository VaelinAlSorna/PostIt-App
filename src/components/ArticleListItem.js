import React from 'react';
//import { IoIosHeart, IoIosChatboxes } from "react-icons/io";
import { Link } from 'react-router-dom';


const ArticleListItem = ({ _id, title, createdAt, creator, comments, likes}) => (
    <Link to={`view/${_id}`}>
        <h3>{title}</h3>
        <p>Posted by: {creator} at: {createdAt}</p>
        {/*<p> <IoIosHeart /> {likes.length}</p>*/}
        {/*<p><IoIosChatboxes /> { comments.length }</p>*/}
    </Link>
);

export default ArticleListItem;