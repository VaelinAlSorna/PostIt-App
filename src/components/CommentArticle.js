import React from 'react';
import uuid from 'uuid';
import moment from 'moment';

class CommentArticle extends React.Component {
    state = {
        comment: '',
        createdAt: moment().format('D. M. Y'),
        error: ''
    };
    onTextareaChange = (e) => {
        const comment = e.target.value;
        this.setState(() => ({
            comment
        }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (this.state.comment.trim().length < 3) {
            this.setState(() => ({
                error: 'Yout comment must contain at least 3 characters.'
            }));
        } else {
            const id = uuid();
            this.props.onSubmit({...this.state, id});
            this.setState(() =>  ({
                comment: '',
                error: ''
            }));
        }
    }
    render () {
        return (
            <form onSubmit={this.onSubmit}>
                { this.state.error && <p>{this.state.error}</p>}
                <textarea 
                    onChange={this.onTextareaChange} 
                    placeholder="What are your thoughts ?"
                    value={this.state.comment}
                >
                </textarea>
                <input type="submit" value="Comment" />
            </form>
        );
    };
};

export default CommentArticle;