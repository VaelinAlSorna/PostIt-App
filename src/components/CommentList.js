import React from 'react';
import CommentListItem from './CommentListItem';

const CommentList = (props) => (
         <ol>
            {
                props.comments.map((comment) => <CommentListItem 
                                                userId={props.userId}
                                                onClickDeleteComment={props.onClickDeleteComment} 
                                                key={comment._id} 
                                                {...comment}/>
                                            )
            }
        </ol>
)

export default CommentList;