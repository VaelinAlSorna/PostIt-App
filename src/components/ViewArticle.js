import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

const ViewArticle = ({ article }) => (
    <div>
        {
            !!article ? (
                <div>
                    <h2>{article.title}</h2>
                    <p>{article.createdAt}</p>
                    <p>{article.text}</p>
                    <Link to={`/edit/${article.id}`}>Edit Article</Link>
                </div>
            ) : (
                <h2> This article doesn't exist! </h2>
            )
        }
    </div>
);

const mapStateToProps = (state, props) =>({
    article: state.articles.find((article) => article.id === props.match.params.id)
});

export default connect(mapStateToProps)(ViewArticle);
