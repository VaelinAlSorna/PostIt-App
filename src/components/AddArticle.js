import React from 'react';
import { connect } from 'react-redux'
import ArticleForm from './ArticleForm';
import { addArticle } from '../actions/articlesActions'


class AddArticle extends React.Component {
    onSubmit = (article) => {
        this.props.dispatch(addArticle(article))
        this.props.history.push('/');
    }
    render () {
        return ( 
        <div>
            <ArticleForm onSubmit={this.onSubmit} />
        </div>
        );
    }
}

export default connect(undefined)(AddArticle);

