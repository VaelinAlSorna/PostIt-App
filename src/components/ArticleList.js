import React from 'react';
import { connect } from 'react-redux';
import ArticleListItem from './ArticleListItem';
import ArticleListFilter from './ArticleListFilter';
import yourArticlesFilter from '../selectors/yourArticlesFilter';

export const ArticleList = (props) => (
    <div className="main-content responsive-scroll">
        <ArticleListFilter />
        <ol className="list">
            {
            !!props.networkError
                ? <p className="server-error"> We are sorry ,but server is not responding ,please try to reconnect later.</p>   
                : (props.articles.length === 0) 
                    ? <p className="no-articles"> No articles fulfill these search conditions. </p>
                    : props.articles.map((article) => <ArticleListItem key={article._id} {...article} />)
            }
        </ol>
    </div>
);
 

const mapStateToProps = (state) => ({
    articles: yourArticlesFilter(state.articles, state.auth.id, state.filter),
    networkError: state.networkError
});

export default connect(mapStateToProps)(ArticleList);