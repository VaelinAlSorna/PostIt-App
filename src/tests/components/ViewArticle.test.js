import React from 'react';
import { shallow } from 'enzyme';
import { ViewArticle } from '../../components/ViewArticle';
import articles from '../fixtures/articlesFixtures';
import user from '../fixtures/usersFixtures';

let wrapper, startAddLike, startRemoveLike, startCommentArticle, startRemoveComment;

beforeEach(() => {
    startAddLike = jest.fn();
    startRemoveLike = jest.fn();
    startCommentArticle = jest.fn();
    startRemoveComment = jest.fn();
    wrapper = shallow(<ViewArticle 
                        article={articles[0]} 
                        startAddLike={startAddLike}
                        startRemoveLike={startRemoveLike}
                        startCommentArticle={startCommentArticle}
                        startRemoveComment={startRemoveComment}
                        userId={user.id}
                    />
                );
});

test('should render ViewArticle correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should render ViewArticle with error message ', () => {
    wrapper = shallow( <ViewArticle />);
    expect(wrapper).toMatchSnapshot();
});

test('should handle startCommentArticle onSubmit', () => {
    wrapper.find('CommentArticle').simulate('submit', articles[0].comments[0]);
    expect(startCommentArticle).toHaveBeenLastCalledWith( articles[0]._id ,articles[0].comments[0]);
});

test('should handle startRemoveComment', () => {
    wrapper.find('CommentList').prop('onClickDeleteComment')(articles[0].comments[0]._id);
    expect(startRemoveComment).toHaveBeenLastCalledWith(articles[0]._id, articles[0].comments[0]._id);
});

test('should handle startAddLike on clicking heart button', () => {
    wrapper = shallow(<ViewArticle 
        article={articles[1]} 
        startAddLike={startAddLike}
        startRemoveLike={startRemoveLike}
        startCommentArticle={startCommentArticle}
        startRemoveComment={startRemoveComment}
        userId={user.id}
        />
    );
    wrapper.find('button').at(1).simulate('click');
    expect(startAddLike).toHaveBeenLastCalledWith(articles[1]._id, user.id)
});

test('should handle startRemoveLike on clicking heart button', () => {
    wrapper.find('button').at(1).simulate('click');
    expect(startRemoveLike).toHaveBeenLastCalledWith(articles[0]._id, user.id)
});


test('editing should be disabled if userId does not match  article._creatorID', () => {
    wrapper = shallow(<ViewArticle 
        article={articles[0]} 
        startAddLike={startAddLike}
        startRemoveLike={startRemoveLike}
        startCommentArticle={startCommentArticle}
        startRemoveComment={startRemoveComment}
        userId={'999'}
        />
    );
    expect(wrapper.find('button').at(0).props().disabled).toBe(true);
});



test('liking should be disabled if user does not exists', () => {
    wrapper = shallow(<ViewArticle 
        article={articles[1]} 
        startAddLike={startAddLike}
        startRemoveLike={startRemoveLike}
        startCommentArticle={startCommentArticle}
        startRemoveComment={startRemoveComment}
        userId={undefined}
        />
    );
    expect(wrapper.find('button').at(0).props().disabled).toBe(true);
});
