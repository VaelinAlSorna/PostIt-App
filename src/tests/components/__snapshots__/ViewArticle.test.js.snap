// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ViewArticle correctly 1`] = `
<div>
  <div>
    <button
      disabled={false}
      onClick={[Function]}
    >
      Edit Article
    </button>
    <h2>
      Breaking news
    </h2>
    <p>
      Posted by:
       at: 
      1.1.1970
    </p>
    <p>
      Political info
    </p>
    <CommentArticle
      likeOrCommentAllowed={false}
      onSubmit={[Function]}
    />
    <button
      disabled={false}
      onClick={[Function]}
    >
       
      <Component
        style={
          Object {
            "color": "grey",
          }
        }
      />
    </button>
    <p>
      Number of Likes: 
      1
    </p>
    <CommentList
      comments={
        Array [
          Object {
            "_id": "1a",
            "comment": "Something to test?",
            "createdAt": "1.1.1970",
          },
        ]
      }
      onClickDeleteComment={[Function]}
      userId="343"
    />
  </div>
</div>
`;

exports[`should render ViewArticle with error message  1`] = `
<div>
  <h2>
     This article doesn't exist! 
  </h2>
</div>
`;
